File: 4. Queue/Double Ended Queue/Input-Restricted Deque/deQueue.c
#include <stdio.h>
#include <stdbool.h> // For using boolean types (true/false)

// Define a constant for the maximum size of the DEQueue.
#define MAX_SIZE 5

// Defines the structure for an Input-Restricted Double-Ended Queue.
// Insertion is restricted to the rear, but deletion is allowed at both ends.
typedef struct {
    int data[MAX_SIZE];
    int front;
    int rear;
} DEQueue;

/**
 * @brief Initializes the DEQueue by setting front and rear to -1.
 * @param dq A pointer to the DEQueue structure.
 */
void initializeQueue(DEQueue *dq) {
    dq->front = -1;
    dq->rear = -1;
}

/**
 * @brief Checks if the DEQueue is empty.
 * @param dq A pointer to the DEQueue structure.
 * @return True if the queue is empty, false otherwise.
 */
bool isEmpty(DEQueue *dq) {
    return dq->front == -1;
}

/**
 * @brief Checks if the DEQueue is full.
 * @param dq A pointer to the DEQueue structure.
 * @return True if the queue is full, false otherwise.
 */
bool isFull(DEQueue *dq) {
    // The queue is full in two scenarios in a circular array:
    // 1. Front is at the beginning (0) and rear is at the end.
    // 2. Rear is just behind the front (circularly).
    return (dq->front == 0 && dq->rear == MAX_SIZE - 1) || (dq->front == dq->rear + 1);
}

/**
 * @brief Adds an element to the rear of the DEQueue.
 * @param dq A pointer to the DEQueue structure.
 */
void enQueue(DEQueue *dq) {
    if (isFull(dq)) {
        printf("\n~~ Queue is Full! ~~\n");
        return;
    }

    int val;
    printf("\nEnter Value:- ");
    scanf("%d", &val);

    // If this is the first element, initialize both front and rear to 0.
    if (isEmpty(dq)) {
        dq->front = 0;
        dq->rear = 0;
    } else {
        // Move the rear pointer forward, wrapping around if necessary.
        if (dq->rear == MAX_SIZE - 1) {
            dq->rear = 0;
        } else {
            dq->rear++;
        }
    }

    // Place the new value at the rear.
    dq->data[dq->rear] = val;

    printf("Insertion Successfull. Value:- %d\n", val);
}

/**
 * @brief Removes an element from either the front or the rear of the DEQueue.
 * @param dq A pointer to the DEQueue structure.
 * @param type 1 to remove from the front, 2 to remove from the rear.
 */
void deQueue(DEQueue *dq, int type) {
    if (isEmpty(dq)) {
        printf("\n~~ Queue is Empty! ~~\n");
        return;
    }

    int toBeDeletedValue;

    // If this is the last element in the queue.
    if (dq->front == dq->rear) {
        toBeDeletedValue = dq->data[dq->front];
        initializeQueue(dq); // Reset the queue to empty.
    } 
    // Remove from the front (type 1).
    else if (type == 1) {
        toBeDeletedValue = dq->data[dq->front];
        // If front is at the end, wrap around to the beginning.
        if (dq->front == MAX_SIZE - 1) {
            dq->front = 0;
        } else {
            // Otherwise, just increment front.
            dq->front++;
        }
    } 
    // Remove from the rear (type 2).
    else if (type == 2) {
        toBeDeletedValue = dq->data[dq->rear];
        // If rear is at the beginning, wrap around to the end.
        if (dq->rear == 0) {
            dq->rear = MAX_SIZE - 1;
        } else {
            // Otherwise, just decrement rear.
            dq->rear--;
        }
    }

    printf("\nData Removed:- %d\n", toBeDeletedValue);
}

/**
 * @brief Shows the element at the front or rear without removing it.
 * @param dq A pointer to the DEQueue structure.
 * @param type 1 for the front element, 2 for the rear element.
 */
void peek(DEQueue *dq, int type) {
    if (isEmpty(dq)) {
        printf("\n~~ Queue is Empty! ~~\n");
        return;
    }

    int peekVal;
    if (type == 1) {
        peekVal = dq->data[dq->front];
        printf("\nFront Element:- %d\n", peekVal);
    } else if (type == 2) {
        peekVal = dq->data[dq->rear];
        printf("\nRear Element:- %d\n", peekVal);
    }
}

/**
 * @brief The main function that drives the program.
 */
int main() {
    int choice;
    // Declare a DEQueue variable (not a pointer).
    DEQueue dq;
    // Pass its address to the functions.
    initializeQueue(&dq);
    printf("Hello! Here You Can Perform Following Input-Restricted DEQueue Operation!");
    while (1) {
        printf("\n1. En-Queue(Rear)\n2. De-Queue Front\n3. De-Queue Rear\n4. Peek Front\n5. Peek Rear\n6. Is Empty\n7. Is Full\n8. Exit\nEnter Your Choice:- ");
        scanf("%d", &choice);

        switch (choice) {
        case 1:
            enQueue(&dq);
            break;
        case 2:
            deQueue(&dq, 1);
            break;
        case 3:
            deQueue(&dq, 2);
            break;
        case 4:
            peek(&dq, 1);
            break;
        case 5:
            peek(&dq, 2);
            break;
        case 6:
            if (isEmpty(&dq)) {
                printf("\n~~ Queue is Empty! ~~\n");
            } else {
                printf("\n~~ Queue is Not Empty! ~~\n");
            }
            break;
        case 7:
            if (isFull(&dq)) {
                printf("\n~~ Queue is Full! ~~\n");
            } else {
                printf("\n~~ Queue is Not Full! ~~\n");
            }
            break;
        case 8:
            printf("\n~~ Thanks For Using! Have a Great Day! ~~\n");
            return 0;
        default:
            printf("Please Enter a Vaild Input!\n");
            break;
        }
    }
    return 0;
}

File: 4. Queue/Double Ended Queue/Input-Restricted Deque/deQueue.js
const prompt = require('prompt-sync')({ sigint: true });

let data = [], front = -1, rear = -1, MAX_SIZE = 5;

function enQueue() {
    if (isFull()) {
        console.log(`\n~~ Queue is Full ~~\n`);
        return;
    }

    console.log("\n");
    let val = Number.parseInt(prompt("Enter Your Number:- "));

    if (isEmpty()) {
        front = rear = 0;
    } else {
        if (rear == MAX_SIZE - 1) {
            rear = 0;
        } else {
            rear++;
        }
    }

    data[rear] = val;

    console.log(`Insertion Successfull:- ${val}\n`);
    return;
}

function deQueue(type) {
    if (isEmpty()) {
        console.log(`\n~~ Queue is Empty ~~\n`);
        return;
    }

    let toBeDeletedData;

    if (front == rear) {
        toBeDeletedData = data[front];
        front = rear = -1; // Reset queue
    } else if (type == "FRONT") {
        toBeDeletedData = data[front];

        if (front == MAX_SIZE - 1) {
            front = 0;
        } else {
            front++;
        }
    } else if (type == "REAR") {
        toBeDeletedData = data[rear];

        if (rear == 0) {
            rear = MAX_SIZE - 1;
        } else {
            rear--;
        }
    }

    console.log(`\nDeletion Successfull:- ${toBeDeletedData}\n`);
    return;
}

function isEmpty() {
    return front == -1;
}

function isFull() {
    return (front == 0 && rear == MAX_SIZE - 1) || (front == rear + 1);
}

function peek(type) {
    if (isEmpty()) {
        console.log(`\n~~ Queue is Empty ~~\n`);
        return;
    }

    let peekVal = 0;

    if (type == "FRONT") {
        peekVal = dq.data[dq.front];
    } else if (type == "REAR") {
        peekVal = dq.data[dq.rear];
    }

    console.log(`\nData Element:- ${peekVal}\n`);
    return;
}


(() => {
    let choice;
    console.log("Hello! Here You Can Perform Following Queue Operation!");
    while (1) {
        console.log("1. En-Queue(Rear)\n2. De-Queue Front \n3. De-Queue Rear\n4. Peek Front\n5. Peek Rear\n6. Is Empty(Checks Queue is Empty)\n7. Is Full(Checks Queue is Full)\n8. Exit");
        choice = Number.parseInt(prompt("Enter Your Choice:- "));
        switch (choice) {
            case 1:
                enQueue();
                break;
            case 2:
                deQueue("FRONT");
                break;
            case 3:
                deQueue("REAR");
                break;
            case 4:
                peek("FRONT");
                break;
            case 5:
                peek("REAR");
                break;
            case 6:
                if (isEmpty()) {
                    console.log(`\n~~ Queue is Empty ~~\n`);
                } else {
                    console.log(`\n~~ Queue is Not Empty ~~\n`);
                }
                break;
            case 7:
                if (isFull()) {
                    console.log(`\n~~ Queue is Full ~~\n`);
                } else {
                    console.log(`\n~~ Queue is Not Full ~~\n`);
                }
                break;
            case 8:
                console.log(`\n~~ Thanks For Using! Have a Great Day! ~~\n`);
                return;
            default:
                console.log(`\n~~ Please Enter a Vaild Input! ~~\n`);
                break;
        }
    }
    return;
})();